Alicerce/Sources/DeepLinking/Route+Tree.swift

```diff
@@ -0,0 +1,270 @@
+
+            var newEdges = edges
+            newEdges[currentComponent.key] = currentComponent.edge(for: childTree)
+
+            return newEdges
+        }
+
+        private func convertLeafToTree(existingHandler: Handler,
+                                       currentComponent: Component,
+                                       remainingRoute: [Component],
+                                       handler: Handler) throws -> ChildEdges {
+            if case .empty = currentComponent {
+                // leaves cannot grow with an already empty component
+                throw Error.duplicateEmptyComponent
+            }
+
+            let newTree = try Tree<Handler>(route: remainingRoute, handler: handler)
+            return [.empty : .simple(.leaf(existingHandler)),
+                    currentComponent.key : currentComponent.edge(for: newTree)]
+        }
+
+        private mutating func removeFromChildTree(_ childTree: Tree<Handler>,
+                                                  edges: ChildEdges,
+                                                  currentComponent: Component,
+                                                  remainingRoute: [Component]) throws -> Handler {
+            var childTree = childTree
+            let handler = try childTree.remove(route: remainingRoute)
+            var newEdges = edges
+
+            // remove the node completely if it became empty (a leaf matched), or update edges with new node
+            if case let .node(nextTreeChildren) = childTree, nextTreeChildren.isEmpty {
+                newEdges[currentComponent.key] = nil
+            } else {
+                newEdges[currentComponent.key] = currentComponent.edge(for: childTree)
+            }
+
+            self = .node(newEdges)
+            return handler
+        }
+
+        private func matchChildEdge(in edges: ChildEdges, component: Component) throws -> Edge? {
+            switch component.key {
+            case .empty:
+                // match empty values separately
+                return edges[.empty]
+            case .constant:
+                // match constants first as their constant value, then match as variables (parameter/wildcard)
+                return edges[component.key] ?? edges[.variable]
+            case .variable:
+                // routes should be matched against empty or constant values only
+                throw Error.invalidComponent(component)
+            }
+        }
+
+        private func matchParameterEdge(parameterName: String?,
+                                        childTree: Tree<Handler>,
+                                        currentComponent: Component,
+                                        remainingRoute: [Component]) throws -> Match {
+            let (parameters, handler) = try childTree.match(route: remainingRoute)
+             
+            guard let parameterName = parameterName else { return (parameters, handler) }
+
+            var allParameters = parameters
```

Comment:
💄 we could do the same as in the `guard`, and redefine the variable: `var parameters = parameters` just an idea 😛